lib/source/defender/aws_iot_defender_api.c

```diff
@@ -551,18 +589,105 @@ void _rejectCallback( void * pArgument,
     AwsIotDefender_Assert( pPublish->u.message.info.pPayload );
 
     /* Invoke user's callback with rejected event. */
+    _handleApplicationCallback( AWS_IOT_DEFENDER_METRICS_REJECTED, pPublish );
+    
+}
+
+/*-----------------------------------------------------------*/
+
+void _mqttAsyncOperationCallback( void * pContext,
+                                  IotMqttCallbackParam_t * pCallbackParams )
+{
+    IotLogInfo( "MQTT Async operation callback" );
+    IotMqttError_t mqttError = pCallbackParams->u.operation.result;
+    IotMqttOperationType_t mqttOp = pCallbackParams->u.operation.type;
+    IotTaskPoolError_t taskPoolError = IOT_TASKPOOL_SUCCESS;
+
+    IotLogDebug( "Operation = %d status = %d\n", mqttOp, mqttError );
+
+    if( mqttError != IOT_MQTT_SUCCESS )
+    {
+        IotLogError( "Operation = %d Failed status = %d\n", mqttOp, mqttError );
+        /* Async operation has failed. Notify application and terminate the defender state machine */
+        _handleApplicationCallback( AWS_IOT_DEFENDER_FAILURE_MQTT, NULL );
+    }
+    else
+    {
+        switch( mqttOp )
+        {
+            case IOT_MQTT_SUBSCRIBE:
+                _subscribed = true;
+                /* Create metrics subscribe job. */
+                taskPoolError = IotTaskPool_CreateJob( _metricsPublishRoutine, NULL, &_metricsPublishJobStorage, &_metricsPublishJob );
+                /* Silence warnigns when asserts are disabled. */
+                ( void ) taskPoolError;
+                AwsIotDefender_Assert( taskPoolError == IOT_TASKPOOL_SUCCESS );
+                /* Schedule Publish Job */
+                taskPoolError = IotTaskPool_Schedule( IOT_SYSTEM_TASKPOOL, _metricsPublishJob, 0 );
+                AwsIotDefender_Assert( taskPoolError == IOT_TASKPOOL_SUCCESS );
+
+                break;
+
+            case IOT_MQTT_PUBLISH_TO_SERVER:
+                taskPoolError = IotTaskPool_CreateJob( _unsubscribeRoutine, NULL, &_unsubscribeJobStorage, &_unsubscribeJob );
+                /* Silence warnigns when asserts are disabled. */
+                ( void ) taskPoolError;
+                AwsIotDefender_Assert( taskPoolError == IOT_TASKPOOL_SUCCESS );
+                IotTaskPool_ScheduleDeferred( IOT_SYSTEM_TASKPOOL,
+                                              _unsubscribeJob,
+                                              _defenderToMilliseconds( AWS_IOT_DEFENDER_WAIT_SERVER_MAX_SECONDS ) );
+                break;
+
+            case IOT_MQTT_UNSUBSCRIBE:
```

Comment:
Constantly modifying subscriptions is significant overhead.  Currently, the sequence of jobs is:

Subscribe -> Publish -> Unsubscribe -> Subscribe -> Publish -> Unsubscribe, etc...

You don't have to unsubscribe if the next thing to do is to resubscribe to the same topic.

In the Defender_Start function, you should make blocking calls to add the subscriptions.  In Defender_Stop, you should make blocking calls to remove the subscriptions.  In between, your Jobs should only do Publishes.